Support for Multiarch Manager Operator (#1417)
openshift/enhancements: enhancements/multi-arch/multiarch-manager-operator.md

## Proposal
Kubernetes 1.27 introduced the scheduling readiness feature[^1]. It
allows users and controllers to set scheduling gates in the
`podSpec.schedulingGates` list at admission time. Gated pods allow mutating some
of the pod spec fields before the pod is scheduled. In particular, the
scheduler's default `Enqueue` plugin holds the pods before entering the
`ActiveQ` queue if a scheduling gate exists in the `podSpec.schedulingGates`
list. Until gates are in the list, the pod will expose its status as
`SchedulingGated`. The controller responsible for a given scheduling gate should
implement the business logic so that the pod can be considered ready to be
scheduled. KEP-3838[^2] introduced the possibility of mutating the node affinity
of pods while they are in `schedulingGated` status (i.e., their
`podSpec.schedulingGates` list has at least one gate).

The multiarch-manager-operator will leverage the scheduling readiness feature
and the mutable pod scheduling directives to update the node affinity of pods
with the set of architectures they support. When a pod is submitted to the API
server, a mutating admission webhook intercepts the pod and adds a scheduling

> gate to the pod spec.

Reviewer: Hey @Prashanth684 and @aleskandro from what I understand here, the Operator will have two components? MutatingAdmissionWebHook and the SchedulingGate? 

With a complicated application, my colleagues in research found the mutating webhook resulted in an unintended out-of-sync status where the changed Pod resource didn't checksum correctly and the AppController ended up reconciling the resource since it wasn't as intended.  As such, maybe there should be some short-circuit methodology or flag in the metadata to skip the mutation.

Thanks for clarifying. Paul

Author: ok. if you can send me the details offline, we can look into it.

  - The pod placement controller will be supported in High Availability (HA)
    mode
  - The webhook will be a stateless component
  - The pod placement controller will support namespace filtering to skip pods
    that should not be mutated
  - No support for customized network scenarios (restricted, proxied,
    disconnected...)
  - No TLS verification for the registries as we are only fetching metadata
  - No caching facilities for the image metadata
  - No graceful uninstallation support: any remaining gated pods have to be
    manually patched to resume scheduling after uninstallation
- Phase 2 - General Availability (GA):
  - support for customized network scenarios (restricted, proxied,
    disconnected...) 
  - implement caching facilities for the image metadata
- Phase 3:
  - TLS verification for the registries
  - Future performance optimizations

### Workflow Description
The following section will highlight the main workflows the operator will be
responsible for from a high-level user perspective.

#### Operator installation
1. The cluster-admin submits a Subscription object to install the
   multiarch-managed-operator, either via CLI or the OperatorHub

> 2. The cluster admin submits the default `PodPlacementConfig` Custom Resource

Reviewer: Have you thought about automatically excluding `openshift-` and `kube-` like the ComplianceOperator?

Author: yes that is the plan.

   (CR) adding their specific configuration related to initializing the
   multiarch-manager-operator's components; the cluster admin can customize the
   CR by excluding namespaces through regexes/globs, a list of configmaps for
   registries that use self signed certs (if any) and customize logging levels
4. The operator reconciles the CR. Its information supports the
   `PodPlacementController` controller and the `SchedulingGateMutatingWebhook`
   mutating admission webhook

#### Pod Admission: gating
1. The user submits a pod
2. The mutating admission webhook is queried and decides whether to add the
   scheduling gate to the pod, based on the namespaces exclusion list
3. The pod is admitted and persisted
4. The pod is put on hold by the scheduler until all the scheduling gates are
   removed

#### NodeAffinity patching of gated pods
1. The `PodPlacementController`, watching for pod events, gets notified of a pod
   event
2. If the multiarch-manager-operator's gate is not present, the pod is ignored
3. If the gate is present, a set of unique images is built from the pod spec

> 4. Each image is inspected to gather the supported architectures set

Reviewer 2: how does the controller choose, locate, and access the correct pull-secret to access the image?

Author: The controller would look at the `ImagePullSecrets` field in the pod spec, lookup the corresponding secret in the pod namespace (which it has access to through the rbac rules) and uses that to access the image. In case this is not present, it falls back to the global pull secret in the openshift-config namespace

Reviewer 2: those aren't the only sources of pullsecrets, the secret can also be chosen simply by looking at what pullsecrets exist in the pod's NS, iirc.

Author: i'll have to look into this but I guess the ImagePullSecret is controlled by the SA - if there is one associated to the pod, or if there is none at all (and no ImagePullSecret explicitly specified by the pod), it should come from the default SA in the namespace.

Author: the behavior has been documented in the latest commit after some prototyping.

permissions:

- Deployment (namespace-scoped)
  - Watch, List, Get, Create, Update, Delete
- Pod (cluster-scoped)
  - Watch, List, Get, Update
- Secrets (cluster-scoped)
  - Watch, List, Get
- ConfigMaps (namespace-scoped)
  - Watch, List, Get
- PodPlacementConfig CRD
  - Watch, List, Get, Update
- Imagestreams/layers (required for internal registry access)
  - Get

#### PodPlacementConfig CR
The configuration CR of the operator allows the admins to set the following:
- a list of `insecureRegistries` to skip the TLS verification;
- a list of `certificateAuthorityRefs` listing configMaps made of key-value
  pairs containing the trusted certificates for the registries that use
  self-signed certificates. For each CA file, the key in the ConfigMap is the
  hostname of the registry in the `hostname[..port]` format, in compliance with
  the `openshift-config/registry-cas` configMap leveraged by the Build APIs for
  similar reasons. By default, the config CR will import the CAs in the above
  mentioned configMap, if any.
- a `skipTLSVerification` flag to completely disable the verification of the TLS
  certificates while inspecting the pods. Although the reduced security, this

>   might guarantee higher performance when patching the nodeAffinity.

Reviewer 3: Hi Ben, the priorities you described drive the phased approach we described.

We would also like this optional operator to be config-driven so that users can leverage it on other Kubernetes distributions. If we want to pre-render and build an automated OCP-specific configuration, we might target it phase by phase and add this implementation to the configuration controller. 

Do you have any suggestions for links (doc/code) to be sure we'll respect the alignment you are claiming wrt CRIO?

>   might guarantee higher performance when patching the nodeAffinity.

Reviewer 2: i suggest you sync w/ @mrunalp or whoever he delegates to ensure alignment w/ kublet/CRIO image pull config+behavior.

>   might guarantee higher performance when patching the nodeAffinity.

Author: after some digging and research, a new section has been added around how we are going to deal with mirror registries and image verification(without requiring explicit configuration to be passed for each of them). I will also consult with Mrunal on this.

>   might guarantee higher performance when patching the nodeAffinity.

Reviewer 4: Yeah, we need to ensure this. cc: @mtrmac and @saschagrunert

>   might guarantee higher performance when patching the nodeAffinity.

Reviewer 5: > Although the reduced security, this might guarantee higher performance when patching the nodeAffinity.

I completely, and in the strongest terms, reject the premise that disabling TLS is a possible way to get acceptable performance.

TLS is what protects users’ pull secrets when contacting registries.

If this feature can not work well enough while enforcing TLS, then we simply don’t have a workable implementation.

(I would also note that I’m skeptical that this is really a major performance concern. To my knowledge every single connection to the API server, and between the API server / etcd, and between etcd nodes, uses TLS, and I’ve never heard disabling TSL would even be considered as an option to improve performance.

And especially with some persistent connections / connection pooling / … I would expect TLS session setup to be pretty negligible. Sure, that might force such features to be moved into phase 1, but again, if this were not performant enough while enforcing TLS, it would simply not be performant enough.

---

If you want to start with a phase 0 that only works with a specific list of public registries like `docker.io` and `quay.io`, and _explicitly refuses_ to use pull secrets (which would, BTW, exclude use of this feature for OCP release content), \*shrug\*, it would be a way to experiment.

But no version of this software that sends users’ pull secrets while disabling TLS can ever be released in the product.

>   might guarantee higher performance when patching the nodeAffinity.

Author: @mtrmac agreed and even we had reservations about doing this but we had thought that with a phased approach we could tackle it later - but makes sense. we will hash out the details and update the proposal shortly. I agree that performance is something we still have to quantify and evaluate before we make sure it is indeed a concern.

>   might guarantee higher performance when patching the nodeAffinity.

Reviewer 5: Thanks. The recent TLS verification updates LGTM.

>   might guarantee higher performance when patching the nodeAffinity.

Author: @bparees could you PTAL the updates? all the concerns should be addressed.

>   might guarantee higher performance when patching the nodeAffinity.

Author: resolving this thread as the concerns have been addressed.

        operations = [CREATE]
        sideEffects = None
        failurePolicy = Ignore
    }

    class PlacementPodConfig {
        <<CR>>

    }

    class Pod {
        schedulingGates: List
        affinity: Object
    }

    class SchedulingGateMutatingWebhook {
        <<Deployment>>
        replicas=3
    }
    SchedulingGateMutatingWebhookSvc --> SchedulingGateMutatingWebhook 

    class SchedulingGateMutatingWebhookSvc {
        <<service>>
    }

    SchedulingGateMutatingWebhookConfiguration --> SchedulingGateMutatingWebhookSvc
    Operator --> PodPlacementController : creates and reconciles
    Operator --> SchedulingGateMutatingWebhook : creates and reconciles
    Operator --> PlacementPodConfig : reconciles
    PodPlacementController --> Pod : updates (patches nodeAffinity,<br/>removes schedulingGate)
    SchedulingGateMutatingWebhook --> Pod : updates (adds schedulingGate)

```
### Open Questions
- What is the performance impact when pods are gated and inspected over many
  namespaces? This needs to be qualified. Mitigations have already been
  discussed in the other sections, but we expect others to address at
  implementation and performance evaluation time.

> - Should the functionality ignore Openshift namespaces (openshift-\*) by

Reviewer 6: The proposed feature should be user workloads only focused. The components under openshift-*, hypershift-* and other openshift dedicated namespaces are owned by OpenShift teams and we know which architectures are supported. By default every architecture is supported unless explicitly state otherwise. Plus, we don't want either of our pods to be mutated in a way that might make the pods unschedulable. Any such situation can potential put the openshift component in risk. The control plane components (etcd, KA, KCM, KS) are static pods so there's no way to mutate those. Nevertheless, openshift-apiserver, oauth-openshift, openshift-controller-manager are not and when improperly mutated they can lead to instability or malfunction of the OpenShift API.

  - Watch, List, Get
- PodPlacementConfig CRD
  - Watch, List, Get, Update
- Imagestreams/layers (required for internal registry access)
  - Get
- operator.openshift.io/ImageContentSourcePolicy
- config.openshift.io/ImageDigestMirrorSet
- config.openshift.io/ImageTagMirrorSet
  - Watch, List, Get
- config.openshift.io/Image
  - Watch, List, Get

#### PodPlacementConfig CR
The configuration CR of the operator allows the admins to set the following:
- a list of `ExcludedNamespaces` listing the namespaces to exclude when running
  the controller logic. Openshift and kube namespaces are excluded implicitly
- a `skipTLSVerification` flag to completely disable the verification of the TLS
  certificates while inspecting the pods. Although the reduced security, this
  might guarantee higher performance when patching the nodeAffinity.

A new Custom Resource Definition (CRD) is being proposed to handle the
configuration of the operator deployment.

```go
// PodPlacementConfigSpec defines the desired state of PodPlacementConfig
type PodPlacementConfigSpec struct {
	// LogVerbosity is the log level for the pod placement controller (1-5)
	// +kubebuilder:default:=1

> 	LogVerbosity int `json:"logVerbosity,omitempty"`

Reviewer: Just a heads up ... You may want to update the CRD to follow this pattern https://github.com/openshift/cluster-kube-descheduler-operator/blob/master/manifests/stable/kube-descheduler-operator.crd.yaml#L62

Author: done.

  of
- `PodPlacementConfigController`: both the `SchedulingGateMutatingWebhook` and
  the `PodPlacementController` consume the configuration information provided by
  admins via the `PodPlacementConfig` CR to mutate the pod and set node affinity
  through image inspection. The `PodPlacementConfigController` is run by the
  operator to roll out the deployments of the business-logic components
- `ImageFacade`: expose an interface hiding the location and nature of the image
  metadata. The `getCompatibleArchitectureSet(image)` method will return the set
  of compatible architectures for a given image. `ImageFacade` uses the
  libraries in `github.com/containers/image`
- `ImageFacadeProxyCache`: it implements an in-memory cache to reduce the
  requests to `ImageFacade` and the external registries following a least
  frequently used cache-eviction policy
- `ImageRegistryController`: it watches the `image.config.openshift.io` and
  adapts the information provided by its singleton object such that the
  registries and image verification related cluster configuration can
  be synced and used by the `RemoteRegistryInspector`

> - `RegistriesMirrorCacher`: it is a controller watching the

Reviewer: Thus the controller would only see one of the supported architectures?

Author: not sure what you mean...this is intended to cover the case where ICSPs are used for mirroring.

Reviewer: Got it

the `SystemConfig` of the image inspector with the information about the
verification of the registries (additional CAs bundles, blocked/allowed
registries, insecure registries...)

### Risks and Mitigations
- The cluster can get into an unrecoverable state due to an issue in the operator:
  - Fail fast policy in the pod placement controller: any error in the pod image inspection is ignored, and the scheduling gate is removed so that the scheduling of the pod is not blocked indefinitely
  - `ignore` failurePolicy for the `SchedulingGateMutatingWebHook`: any non-recognized error raised when the kube-apiserver tries to reach the Webhook is ignored so that the pods can continue to be admitted
- Performance of pod scheduling when pods are gated, and images are inspected for a many pods. A possible mitigation being considered is the caching of the image inspection results and searching the cache first before inspecting the image.
  - Concurrent reconciliation and image inspection: the nature of the job the pod placement controller will run is implemented as a strongly concurrent task leveraging both concurrent reconciliation of gated pods and loop-level parallelism

>   - Namespace-level sharding: future implementations of this operator for strongly loaded clusters can make use of a per-namespace deployment of the controller such that the load of reconciling pods is balanced across namespaces/shards of the cluster

Reviewer: Does your approach match the ImageRegistry? e.g., on a specific Node. maybe it should be deployed to the node with the ImageRegistry pods?

Author: why? what would be the benefit of that ?

Reviewer 3: Hi Paul, if I understand what you say, it's a good point that we discussed: to minimize the inter-pod (and node) network latencies concerning this operator's components dependencies. However, those considerations need further discussions, and we'd get this research and optimizations as a later step, especially once we can start getting data points about this operator's performance and impact on large clusters.

Author: correct. Even then there is no guarantee that Kubernetes will choose based on locale and with images that point to registries that need network access, it will not help anyway. But as @aleskandro said, the solution will evolve based on our findings.

gate to the pod spec.

The `PodPlacementController` watches the pod resource. Suppose any pod has the
scheduling gate set. In that case, the images in the pod spec are inspected, a
subset of architectures that satisfy all containers in the pod is built, and the
pod spec is patched with the additional node affinity constraints. If any
existing node selector requirements are set with key `kubernetes.io/arch`,
the `PodPlacementController` does not override the behavior previously
pushed by other stakeholders, leaving it as complies with KEP-3838[^2].

The multiarch-manager-operator will be available as part of the OperatorHub, as
a secondary operator managed by the Operator Lifecycle Manager (OLM). It will be
a cluster-scoped operator that only cluster admins can install and configure.

Its development is expected to ship in the following phases:

- Phase 1 - Technology Preview (TP):
  - The pod placement controller will be supported in High Availability (HA)
    mode
  - The webhook will be a stateless component
  - The pod placement controller will support namespace filtering to skip pods
    that should not be mutated

>   - No support for customized network scenarios (restricted, proxied,

Reviewer 4: We should plan to use c/image libraries so we can add support for other scenarios more easily in the future.

Reviewer 2: @Prashanth684 is this suggestion encoded in the EP?  can it be resolved?

Author: yes it is. the implementation details [here](https://github.com/openshift/enhancements/pull/1417/files#diff-4fdb9132744e05eb241500caf281668871c3b14b3ebd6eee8722774e07cbd33eR385) mentions  using the c/image libs.

    disconnected...)

>   - No TLS verification for the registries as we are only fetching metadata

Reviewer 5: … and sending pull secrets, no?

Exposing users’ pull secrets is unacceptable even in a prototype, because there’s a high risk the future phases will simply not happen (partially or at all), and the software will be deployed to production in that state. Once a valuable software exists in any shape, there is high pressure to make it available soon.

TLS **must** work correctly from the very start.

Author: agree. we had not considered that the pull secrets are also exposed over the network.

Reviewer 3: Hi @mtrmac, we have addressed all comments regarding the TLS verification in the last commits. We'd like to keep the conversation to only one of the opened threads (i'd say the Ben's one above where you summarized the requested change).

Reviewer 5: Thanks. The recent changes LGTM.

  - No caching facilities for the image metadata

>   - No graceful uninstallation support: any remaining gated pods have to be

Author: Yes,  as described in the risks and mitigations section, the pod would need to be manually patched to remove the scheduling gate to continue with normal scheduling. We might tackle this in later phases - the problem here is that when the operator is uninstalled OLM does not have a concept of ordered deletion. So it would remove the RBAC rules needed to access the pods and the controller would then fail when it tries to get a list of pods to remove the scheduling gate before exiting.

>   - No graceful uninstallation support: any remaining gated pods have to be

Author: @joelanford do you know if there is a way that OLM operators do this sort of cleanup?

>   - No graceful uninstallation support: any remaining gated pods have to be

Reviewer 7: Oh right, yeah, sounds like you need some way to uninstall things, but that would probably mean somehow using finalizers to make sure you have all the rbac/resources required to clean up, and, well, it's going to get messy isn't it 🤔

>   - No graceful uninstallation support: any remaining gated pods have to be

Author: that is correct. even if we use finalizers on the controller - the rbac rules will be cleaned up. not sure if there is a way to set finalizers for rbac and other resources.

>   - No graceful uninstallation support: any remaining gated pods have to be

Reviewer 7: You should be able to set a finalizer on any object, the field is present in the API for every object in Kube so I'd expect it to work

>   - No graceful uninstallation support: any remaining gated pods have to be

Author: we will have it documented and yes, should be simple enough to provide a script. i updated the risks and mitigations section with this detail.

>   - No graceful uninstallation support: any remaining gated pods have to be

Reviewer 2: @Prashanth684 can this conversation be resolved?

    manually patched to resume scheduling after uninstallation
- Phase 2 - General Availability (GA):
  - support for customized network scenarios (restricted, proxied,
    disconnected...)
  - support mirror registries through ICSPs/IDMSs
  - implement caching facilities for the image metadata
- Phase 3:

>   - TLS verification for the registries

Reviewer 5: This must exist from the very start, not as a low priority item that is the first thing to cut.

#### Variation
- *Pods ignored during scheduling*: the `PodPlacementConfig` CR allows a
  namespace exclusion list. Any pods belonging to the excluded namespaces will
  be gated at admission time and will be ignored by the `PodPlacementController`
- *NodeAffinity block already set in the gated pod*: when patching the node
  affinity, any node selector term having a `matchExpression` element using the
  key `kubernetes.io/arch` is preserved. For all the other node
  selector terms, we append the architecture-specific `matchExpression`.
- *Image inspection failure*: in the node affinity patching workflow, if there
  are any errors during the image inspection process (authentication issues,
  network errors, ...), the scheduling gate is removed, and no changes are made
  to the pod. In this case, the scheduling will only be architecture-aware if
  the user submitting the pod defines the affinity predicates explicitly.

### API Extensions

#### Mutating Admission Webhook and Scheduling Gate 
The multiarch-manager-operator reconciles not-yet-scheduled pods using the
scheduling readiness[^1] and pod mutable scheduling directives[^2] features.

> Since Kubernetes 1.27[^3] promoted those features to beta, no additional changes

Author: the [pod scheduling readiness](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3521-pod-scheduling-readiness) and[ pod mutable scheduling directives](https://github.com/kubernetes/enhancements/blob/ac65b5853a9a5b8fb77a8c3069c07780057a4c9e/keps/sig-scheduling/3838-pod-mutable-scheduling-directives/README.md) KEPs were implemented and graduated to beta in 1.27, so the required feature is already in.

> Since Kubernetes 1.27[^3] promoted those features to beta, no additional changes

Author: Is there no precedence in Openshift where we have GA'ed a feature with a beta API? Is that a possibility? Looking at https://kubernetes.io/blog/2020/08/21/moving-forward-from-beta/, it seems there is a possibility that a feature could be stuck in beta.. we are going to work upstream with the sig-scheduling folks to understand what it would take to GA this API.

> Since Kubernetes 1.27[^3] promoted those features to beta, no additional changes

Reviewer 8: > Is there no precedence in Openshift where we have GA'ed a feature with a beta API?

Just because we've made a mistake in the past does not mean we must make a mistake in the future.  Determining which beta features we're willing to depend upon requires consideration of the likelihood of progression, whether we are investing in promoting that feature, and other upstream consideration.  This particular one I'm not familiar with and I suggest you contact @soltysh about discussing on the workloads arch call.

> Since Kubernetes 1.27[^3] promoted those features to beta, no additional changes

Author: yes i'm not saying we should include this indiscriminately. I'm just worried about the feature getting stuck in beta.

to the Openshift Cluster Operators are required to leverage them in the context
of automatically setting the architecture-related node affinity of pods. As
described in the Workflow Description section, the
`SchedulingGateMutatingAdmissionWebhook` changes a pod at admission time. It
adds the `multi-arch.openshift.io/arch-aware-placement` scheduling gate to newly
created pods. The scheduling gate can only be added to pods at creation time.
The `PodPlacementController` is responsible for deleting the scheduling gate

> from the pod after inspecting the architecture of the pod containers' images and

Author: we have proposed fail open now - just to keep the behavior equivalent to not having the operator installed at all. If the image inspection fails, the pod is updated with the gate removed  in which case it might be scheduled on a node where the image cannot run and crashloopbackoff - the reason being that the pod can always be manually patched with the correct affinities. I am not sure if there are arguments for blocking the scheduling by keeping the gate and failing in which case manual intervention would be required as well. Such would be the case if a pod absolutely should not land on the inappropriate arch node for some reason.

> from the pod after inspecting the architecture of the pod containers' images and

Reviewer 3: Some are already at https://github.com/openshift/enhancements/pull/1417/files#diff-4fdb9132744e05eb241500caf281668871c3b14b3ebd6eee8722774e07cbd33eR546 . We will make it clearer.

> from the pod after inspecting the architecture of the pod containers' images and

Reviewer 3: Clarified on the last commit.

patching the node affinity set in the pod. The controller inspecting the image
has access to any secrets in the cluster and builds the credentials to access
the registries consuming the inspected pod's `imagePullSecrets` field. The
controller will also read the image content source policy objects once the
support for restricted network scenarios are implemented. 

#### Required RBAC permissions

The service account running the operator components will need the following RBAC
permissions:

- Deployment (namespace-scoped)
  - Watch, List, Get, Create, Update, Delete
- Pod (cluster-scoped)
  - Watch, List, Get, Update
- Secrets (cluster-scoped)
  - Watch, List, Get
- ConfigMaps (namespace-scoped)
  - Watch, List, Get
- PodPlacementConfig CRD
  - Watch, List, Get, Update
- Imagestreams/layers (required for internal registry access)
  - Get
- operator.openshift.io/ImageContentSourcePolicy
- config.openshift.io/ImageDigestMirrorSet
- config.openshift.io/ImageTagMirrorSet
  - Watch, List, Get
- config.openshift.io/Image
  - Watch, List, Get

#### PodPlacementConfig CR
The configuration CR of the operator allows the admins to set the following:

> - a list of `ExcludedNamespaces` listing the namespaces to exclude when running

Reviewer 3: This was initially thought for excluding the core namespaces which did not seem to expose common patterns helpful for a matchExpression. However, we might consider both cases. 

cc @Prashanth684

  the controller logic. Openshift and kube namespaces are excluded implicitly

> - a `skipTLSVerification` flag to completely disable the verification of the TLS

Reviewer 5: I don’t think this should exist at all. Users should not be asked to manage TLS enforcement for the same registry in two places.

https://github.com/openshift/api/blob/b2e6ab3d6c69e3efb4e7d52184b5794b8359165b/config/v1/types_image.go#L115 already exists, so I would expect there to be _some_ way for this software to consume that value.

It doesn’t make much sense to me to introduce a new CR with a new field and to immediately plan to deprecate it. Just read the right data from the start.

  certificates while inspecting the pods. Although the reduced security, this
  might guarantee higher performance when patching the nodeAffinity.

A new Custom Resource Definition (CRD) is being proposed to handle the
configuration of the operator deployment.

```go
// PodPlacementConfigSpec defines the desired state of PodPlacementConfig
type PodPlacementConfigSpec struct {
	// LogVerbosity is the log level for the pod placement controller
  //
	// Valid values are: "Normal", "Debug", "Trace", "TraceAll".
	// Defaults to "Normal".
	// +optional
	// +kubebuilder:default=Normal

> 	LogVerbosity int `json:"logVerbosity,omitempty"`

Author: typo. thanks for catching.

	// ExcludedNamespaces is the list of namespaces the pod placement controller should ignore
	// +kubebuilder:default:={openshift-*,kube-*,hypershift-*}

> 	ExcludedNamespaces []string `json:"excludedNamespaces,omitempty"`

Author: > Also, if this is defaulted to those values, what happens if a user removes one of the values? Is that an issue? Wondering if instead the operator should make sure these values are always excluded

Yes, we are thinking of implicitly excluding these namespaces and not even having a default.

> Do you expect this list to evolve over time? I'm wondering if this would be considered a configuration API in openshift terms. If it is, we may want to default this controller side so that we can change the list later at the discretion of the controller itself.

Yes that is the plan. The API is part of the operator/controller itself. Is that the norm of optional operators?

> 	ExcludedNamespaces []string `json:"excludedNamespaces,omitempty"`

Reviewer 7: Whether it's the norm or not I don't know, but I think you can follow the same patterns that core openshift operators do here. We have a standard prose that we put in for defaults, but, as this isn't necessarily a default, it may differ slightly. I'm thinking the doc would read something like:

```
// When omitted, this means no opinion and the operator is left to choose a reasonable default.
// The current default behaviour is to ignore only reserved namespaces.
// Reserved namespaces are those matching the following globs; openshift-*,kube-*,hypershift-*.
// The list of reserved namespaces will always be excluded, independent of the value of this field.
// The list of reserved namespaces is subject to change over time.
```

> 	ExcludedNamespaces []string `json:"excludedNamespaces,omitempty"`

Reviewer 3: added the doc comment and will report it in the poc code too.

    // +kubebuilder:default:=true

> 	SkipTLSVerification bool `json:"skipTLSVerification,omitempty"`

Reviewer 3: As per the discussions above, the SkipTLSVerification field has been dropped at all. We will rely on the fields available in the image registry config CR.

Kubernetes Pod Admission controllers through the ones associated with the
service account (SA) running the pod.

##### Registry reachability (proxy and mirroring)
In phase 2, the `RegistriesMirrorCacher` will watch ICSPs/IDMSs to sync the
registry mirror configurations within a map having a `source` as key and the
list of `mirrors` as values.  The inspection of the images will change such that
if the map has keys available, before trying to reach the original URI, the URI
will be looked up in the map (the registry base URL only). If a hit occurs, the
mirrors will be tried in order to inspect the image until the inspection
succeeds or the list is fully visited. Finally, no changes are expected to be
applied to support configurations like cluster-wide proxy at the time of
writing.

##### Registry and image verification
While authenticating is critical to inspect images, so is the verification of
the images and the registry being inspected. However, this is given lower
priority as the operator is only inferring metadata from the image as opposed to
pulling it.

In Phase 1, the `skipTlsVerification` field is proposed in the CR and is set to

> true by default. It ignores verification of the TLS certificates the registry

Reviewer 5: Exposing users’ credentials on the network is **completely unacceptable**.

Any version that leaves the original developers’ workstation simply must not default to disabling TLS. It would be irresponsible to include any code with that behavior into a shipped product.

provides with the trusted CAs bundle. In phase 2, we will introduce the
`ImageRegistryController`, watching the `image.config.openshift.io/cluster`
CR[^7]. The `imageRegistryController` will cache and adapt the image registries
configuration stored in the `image.config.openshift.io/cluster` object, to set
the `SystemConfig` of the image inspector with the information about the
verification of the registries (additional CAs bundles, blocked/allowed
registries, insecure registries...)

### Risks and Mitigations
- The cluster can get into an unrecoverable state due to an issue in the operator:

>   - Fail fast policy in the pod placement controller: any error in the pod image inspection is ignored, and the scheduling gate is removed so that the scheduling of the pod is not blocked indefinitely

Author: yes like I said in the other comment - this would be good to have if there is a use case where a pod should not be scheduled on a certain node due to some catastrophic reasons, but we haven't seen a use case like that - which is why any failure of the controller would make pods behave like they normally do - without this feature existing at all. the pods would crashloopbackoff is scheduled  where they cannot run.

>   - Fail fast policy in the pod placement controller: any error in the pod image inspection is ignored, and the scheduling gate is removed so that the scheduling of the pod is not blocked indefinitely

Reviewer 3: > Is it likely a user might want to have configuration over this behaviour?

any nodeSelectorTerms with the same key (kubernetes.io/arch) defined by the users will not be overwritten, and the controller will skip patching them following the rules reported in https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3838-pod-mutable-scheduling-directives#design-details, similar to what we're doing in the PoC at https://github.com/aleskandro/multiarch-operator/blob/master/controllers/pod_reconciler.go#L132-L175.

    performance. Note that this is more of a scheduling metric than a metric
    of the multiarch-manager-operator.
- SLOs will be based on the metrics above and will be defined after performance
  testing and evaluation: in general, we expect the SLO on the webhook to be
  well defined, unlike the pod placement controller, because the latter will
  need to account for the reliability of the external registries it communicates
  with over the network
- Alerts
  - webhook service/pods unreachable/reporting errors for more than X minutes
  - Controller pods health in a given time window
  - Gated pods per second / scheduled pods > 1: not 100% guaranteed to be
      operator-related
- Impact
  - A certain duration of time will be added to every pod which is intercepted
    and inspected
  - Webhook has FailPolicy=Ignore which will not block pods from scheduling on
    webhook errors

#### Failure Modes
- Webhook failure - The mutating admission webhook has the "FailPolicy=Ignore"
  setting. The creation or scheduling of pods will not be blocked if the webhook
  is down. However, there would be an alert for this event notifying the
  administrator about this condition.

> - Operator/controller failure - Any operator/controller failure will be

Author: that would likely be taken care of by implicitly ignoring openshift namespaces (yes, we have to make sure this operator is deployed as openshift-*)

> - Operator/controller failure - Any operator/controller failure will be

Reviewer 3: We might also leverage the matchExpression-like logic (with a default) to exclude it from being considered (related another comment above)

#### Pod Admission: gating
1. The user submits a pod
2. The mutating admission webhook is queried and decides whether to add the
   scheduling gate to the pod, based on the namespaces exclusion list
3. The pod is admitted and persisted
4. The pod is put on hold by the scheduler until all the scheduling gates are
   removed

#### NodeAffinity patching of gated pods
1. The `PodPlacementController`, watching for pod events, gets notified of a pod
   event
2. If the multiarch-manager-operator's gate is not present, the pod is ignored
3. If the gate is present, a set of unique images is built from the pod spec
4. Each image is inspected to gather the supported architectures set
5. The intersection of the supported architecture sets builds the pod compatible
   architecture set
6. If no inspection errors occurred, the
   `podSpec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution`
   stanza is patched with the architecture node selector requirements. If any
   node selector term is present, they are all patched. Otherwise, a new node
   selector term is added to the `.matchExpressions` list. The merge strategy
   described here follows the rules in KEP-3838[^2]

> 7. The scheduling gate is removed from the pod spec and the pod is updated

Author: I can add the concern around leader election latency to the open questions. One way to mitigate this would be to setup the operator manager with ` LeaderElectionReleaseOnCancel` since we are not planning any cleanup (yet)

> 7. The scheduling gate is removed from the pod spec and the pod is updated

Author: added this concern to the open questions.

            PodPlacementController ->> PodPlacementController: Intersect Supported Architectures
            PodPlacementController ->> Pod: PatchNodeAffinity
            Note over PodPlacementController: inspection errors are ignored<br>to allow the removal of the scheduling<br>gate.
            deactivate PodPlacementController
            PodPlacementController ->> Pod: RemoveSchedulingGate
        end
    end
    Note over Scheduler: no more schedulingGates
    Scheduler ->> Scheduler: move Pod to activeQ <br> and other plugins

```
#### Deletion of the CR
1. The cluster admin deletes the CR
2. All resources related to the operator - the controllers, webhook service and
   pods automatically get deleted
3. Any pods stuck in the scheduling gated phase will have to be patched manually
   to resume scheduling (by removing the scheduling gate)
4. Recreating the CR again will install all related resources

#### Operator uninstallation
- Operator uninstallation removes the operator. If the option to uninstall
  operands is selected, all resources are removed. 

#### Variation
- *Pods ignored during scheduling*: the `PodPlacementConfig` CR allows a

>   namespace exclusion list. Any pods belonging to the excluded namespaces will

Author: oops..fixed.

  proxy/cluster's `spec.trustedCA` config map reference and the global CA bundle deployed in the nodes. It is mounted in
  the controller pod at `/etc/pki/tls/certs/ca-bundle.crt` so that the crypto Golang package will source it[^9] to
  perform the TLS handshake and verification. The operator watches the ConfigMap and redeploys the Pod placement
  controller if it changes.
- Additional CA bundles: these are inferred from the configmap `image-registry-certificates` in
  the `openshift-image-registry` namespace, to track any additional CA bundles set in
  the `image.config.openshift.io/cluster` object and only used to verify image registries defined within the ConfigMap.
  The pod placement controller watches this config map for updates and patches the `/etc/docker/certs.d` directory
  according to changes in the ConfigMap, similar to the node-ca DaemonSet managed by the image registry operator[^9].
- Finally, the Pod Placement Controller watches the `image.config.openshift.io/cluster`[^7] and patches
  the`/etc/containers/policy.json` and `/etc/containers/registries.conf` files in the pod placement controller according

>   to the settings of the `image.config.openshift.io/cluster` object's registrySources fields

Reviewer 5: (BTW callers of the containers/image library can point it at other locations, so the code does not necessarily need root inside the container to be able to overwrite files in `/etc`.)

Reviewer 3: We will not leverage a privileged/rooted pod for changing this folder and will address the actual location of the files at implementation time.

image binaries can be executed. This strongly impacts the user experience: it
couples the set of compatible architectures at a given point in time to the pod
spec, does not scale well, and is challenging to maintain. The
`multiarch-manager-operator` aims to automate the inspection of the container
images, derive a set of architectures supported by a pod and use it to
automatically define strong predicates based on the `kubernetes.io/arch`
label in the pod's nodeAffinity.

## Motivation
- Easy adoption of multi-architecture compute clusters by customers with varying
  workload images which might not be built for specific architectures
- Provides users a "clean" user experience without having to worry about
  manually modifying deployment specs when deploying workloads on a multi-arch
  compute clusters

### User Stories
- As a user who wants to deploy workloads on a multi-arch compute cluster, I
  want the pods to be scheduled on the appropriate subset of nodes based on the
  architectures supported by their containers' images so that the coupling
  between the characteristics of the images and the nodeAffinity declared in

>   the pod spec template does not require future manual maintenance due to image

Reviewer 2: @Prashanth684 can you resolve this comment?

  updates provided by the external stake-holders maintaining them
- As a cluster administrator, when I add a new node of a secondary architecture
  to a cluster, I do not want to taint the nodes to be unschedulable to ensure
  that workloads continue to deploy and run as expected given that the images
  used for deploying workloads may not support the above said secondary
  architecture
- As a cluster administrator who wants to migrate from a single-arch to a
  multi-arch compute cluster, I do not want to rebuild each container image for
  all the architectures that are present in the cluster
- As a user who wants to deploy workloads on a multi-arch compute cluster, I do
  not want to modify every pod (or any resources owning pods) spec to include a
  node affinity predicate to filter the nodes at scheduling time, based on the

>   compatible architectures

Reviewer 3: Any node selector/affinity about the node architecture manually set by the users is prioritized: the controller will not overwrite those predicates externally set by the users.

The requests/limits concerns are interesting and worth a broader discussion IMO. However, I believe they are out of the scope of this EP.

>   compatible architectures

Author: This is a good point and needs to be covered in documentation letting the user know that their pod can be scheduled on nodes with any of the architectures matching the images. This would then require manual intervention from the user to either exclude the namespaces where such pods are deployed or specify their own node selector arch preference.

>   compatible architectures

Author: added this concern to the risks and mitigations section

pushed by other stakeholders, leaving it as complies with KEP-3838[^2].

The multiarch-manager-operator will be available as part of the OperatorHub, as
a secondary operator managed by the Operator Lifecycle Manager (OLM). It will be
a cluster-scoped operator that only cluster admins can install and configure.

Its development is expected to ship in the following phases:

- Phase 1 - Technology Preview (TP):
  - The pod placement controller will be supported in High Availability (HA)
    mode
  - The webhook will be a stateless component
  - The pod placement controller will support namespace filtering to skip pods
    that should not be mutated
  - No support for customized network scenarios (restricted, proxied,
    disconnected...)
  - No caching facilities for the image metadata
  - No graceful uninstallation support: any remaining gated pods have to be
    manually patched to resume scheduling after uninstallation
- Phase 2 - General Availability (GA):
  - support for customized network scenarios (restricted, proxied,
    disconnected...)
  - support mirror registries through ICSPs/IDMSs
  - implement caching facilities for the image metadata
- Phase 3:

>   - Future performance optimizations

Reviewer 2: @Prashanth684 this should be discussed under risks/drawbacks and/or testing.

>   - Future performance optimizations

Reviewer 3: Hi @bparees, it is already highlighted in both the risks/drawbacks, testing and open questions sections.

>   - Future performance optimizations

Author: resolving this thread.

### Workflow Description
The following section will highlight the main workflows the operator will be
responsible for from a high-level user perspective.

#### Operator installation
1. The cluster-admin submits a Subscription object to install the
   multiarch-manager-operator, either via CLI or the OperatorHub
2. The cluster admin submits the default `PodPlacementConfig` Custom Resource
   (CR) adding their specific configuration related to initializing the
   multiarch-manager-operator's components; the cluster admin can customize the
   CR by using a namespace selector to control pod filtering and logging levels
4. The operator reconciles the CR. Its information supports the
   `PodPlacementController` controller and the `SchedulingGateMutatingWebhook`
   mutating admission webhook

#### Pod Admission: gating
1. The user submits a pod

> 2. The mutating admission webhook is queried and decides whether to add the

Reviewer 3: Hi @ingvagabund, as far as we know, there shouldn't be an issue for other schedulers unless they are based on older Kubernetes APIs. 

In such cases of schedulers built on top of older k8s APIs than the running ones (>=1.27), the scheduler will not know about the schedulingReadiness feature. Still, it will not schedule the pod in any case as the DefaultBinder plugin will get an error from the kube-apiserver, and fail until the schedulingGates are removed from our controller.

Moreover, shouldn't we recommend the users not use a custom-scheduler built on top of API versions that are non-compatible with the Kubernetes in their cluster?

Are these the cases you are concerned about?

> 2. The mutating admission webhook is queried and decides whether to add the

Reviewer 6: I was stressing the case where two different schedulers operate over the same namespace. Where a user is interested in mutating only pods scheduled by one of the schedulers, ignoring pods of the other one.

One can always generate pods that do not have two or more different values for .spec.schedulerName inside the same namespaces. So, this is not something critical to deal with now. Just another use case that might be interesting for users.

> 2. The mutating admission webhook is queried and decides whether to add the

Reviewer 3: It seems a good point to me, too @ingvagabund.

I wonder if we can use a similar approach to the namespace selectors by providing a pod selector.
We wouldn't bind to the schedulerName, but provide a second-level, more fine-grained approach to include or exclude pods based on their labels.

I'd add this as an open question.

            PodPlacementController ->> PodPlacementController: Intersect Supported Architectures
            PodPlacementController ->> Pod: PatchNodeAffinity
            Note over PodPlacementController: inspection errors are ignored<br>to allow the removal of the scheduling<br>gate.
            deactivate PodPlacementController
            PodPlacementController ->> Pod: RemoveSchedulingGate
        end
    end
    Note over Scheduler: no more schedulingGates
    Scheduler ->> Scheduler: move Pod to activeQ <br> and other plugins

```
#### Deletion of the CR
1. The cluster admin deletes the CR
2. All resources related to the operator - the controllers, webhook service and
   pods automatically get deleted
3. Any pods stuck in the scheduling gated phase will be patched without changes to
   the node affinity and their scheduling will ungate
4. Recreating the CR again will install all related resources

#### Operator uninstallation
- Operator uninstallation removes the operator. If the option to uninstall
  operands is selected, all resources are removed.
- Any pods stuck in the scheduling gated phase could need a manual patch

>   performed by the users to remove the scheduling gate and resume scheduling

Author: added this to the new documentation plan section.

>   performed by the users to remove the scheduling gate and resume scheduling

Reviewer 2: above it says deleting the CR will delete the controllers, so how does a use "remove the CR, then wait until scheduling gates are removed before uninstalling the operator"?

uninstalling the operator actually has no effect on the controller(operand) anyway, unless you choose delete operands, in which case that's the same as deleting the CR (where, again, it's not clear how a user can choose to wait before deleting the controller).

>   performed by the users to remove the scheduling gate and resume scheduling

Author: There are a few cases in terms of deletion of the operator:
- uninstall operator
-  uninstall operator + operands - pods might be left in scheduling gated state which would need to be manually patched . We noted this in the risks/documentation section where we would need to be explicit and provide clear guidance to the user on this case.
- delete CR + uninstall operator - this is essentially the same as the above case, but deleting the CR first would trigger the controller to start patching the pods which are scheduling gated manually. In case the user prematurely removes the operator (by also selecting remove operands), pods will be stuck in scheduling gated state which would have to be removed manually.

Is it also possible to not expose the option for a particular operator to remove operands. i.e, not make that check box available? in that case the uninstall option would only remove the operator, but the cleanup of operand resources would have to be manual if the operator is removed before the CR.

>   performed by the users to remove the scheduling gate and resume scheduling

Author: In any case, the plan for now is to allow uninstallation of the operator (with operands) in which case the pods which are scheduling gated will have to be patched manually - this will be a documentation effort. The other method is deletion of the CR which will un-gate any scheduling gated pods provided the user gives sufficient time by monitoring these pods to see if the gate has been removed. These points have been captured in this proposal.

>   performed by the users to remove the scheduling gate and resume scheduling

Reviewer 9: > Is it also possible to not expose the option for a particular operator to remove operands. i.e, not make that check box available?

This is actually possible. See https://github.com/openshift/console/pull/12234

You would put the following annotation in the CSV:
```yaml
"console.openshift.io/disable-operand-delete": "true"
```

Hopefully the annotation makes clear that this is only something that console takes action on. Other clients (e.g. oc, kubectl) and OLM do not do anything to prevent operand deletion coinciding with the operator uninstall.

>   performed by the users to remove the scheduling gate and resume scheduling

Reviewer 2: > Hopefully the annotation makes clear that this is only something that console takes action on. Other clients (e.g. oc, kubectl) and OLM do not do anything to prevent operand deletion coinciding with the operator uninstall.

but they(oc, kubectl, other clients besides console) also don't implement anything to delete(or allow the user to request deletion of) the operands during operator uninstall, right?  Not that someone couldn't create a client that does, of course.

>   performed by the users to remove the scheduling gate and resume scheduling

Reviewer 9: Correct. Console has some special logic (via some vendored `kubectl-operator` plugin code) to list the operands for a particular operator, and then perform a semi-orderly deletion.

Other clients are on their own to deal with deleting operands, CSVs, subscriptions, RBAC, etc.

>   performed by the users to remove the scheduling gate and resume scheduling

Author: Having this option disabled will prevent the user from deleting operands when the scheduling gates are being removed (triggered through CR deletion). Of course, the user could always force delete through oc, but coupled with good documentation - this can be mitigated.

The scheduling gate can only be added to pods at creation time.
The `PodPlacementController` is responsible for deleting the scheduling gate
from the pod after inspecting the architecture of the pod containers' images and
patching the node affinity set in the pod. The controller inspecting the image
has access to any secrets in the cluster and builds the credentials to access
the registries consuming the inspected pod's `imagePullSecrets` field. The
controller will also read the image content source policy objects once the
support for restricted network scenarios are implemented.
If any unrecoverable error occurs during the image inspection process,
the scheduling gate is removed from the pod, and no changes are made to the pod's
affinity.
This logic prevents the cluster's workloads from being indefinitely held
before entering the ActiveQ scheduling queue.

#### Required RBAC permissions

The service account running the operator components will need the following RBAC
permissions:

- Deployment (namespace-scoped)
  - Watch, List, Get, Create, Update, Delete
- Pod (cluster-scoped)
  - Watch, List, Get, Update
- Secrets (cluster-scoped)

>   - Watch, List, Get

Reviewer 7: This is likely to raise some concerns in a security review, can you add a note here as to why this is necessary

Is there anything we can do to mitigate this?

Author: I am not sure there is a way to mitigate this.

Essentially, access to the global pullsecret and the namespace specific pull secret is necessary to consolidate them and write them to a temp file which is then provided to the c/Image libs API to access the image in the registry to get the arch metadata info. I will add this note.

Author: Also what concerns might it raise? Is there some best practices around secrets access?

Reviewer 2: > Also what concerns might it raise? Is there some best practices around secrets access?

i was also going to comment on this as a potential concern but didn't since it seems like a requirement to implement the functionality.

but yes, the concern would around any component that has this level of RBAC because it becomes a target for exploiting.  If there is a way to compromise this component, it becomes a way to leverage higher access into the cluster by reading cluster+cloud credentials.

During implementation it will be important to ensure there's no way to leak the secrets through logs, filesystem, or confused deputy type attacks.

but ultimately this is an add-on operator and cluster-admins will have to decide if they are willing to accept the permissions to use the operator.

Reviewer 3: We have spread some mitigations to limit the exploitation surface of the secrets read-only access throughout the EP; others are considered implementation or deployment details:
1. Leveraging basic software engineering principles (interface segregation, encapsulation...) to limit the exposed code surface capable of accessing secrets.
2. Separating the webhook pods (exposing a network service) and those running the controller (only communicating with the kube-proxy and requesting access to the secrets).
3. Leveraging the least privilege principle by using different service accounts such that (for this specific concern) only the required component can access the secrets globally.
4. Limiting the pods' capabilities via strict security contexts and minimizing the container images running the components.

adding this discussion in the risks/mitigation section.

Author: yes we definitely need to keep these mind when implementing and as @aleskandro mentions above, using different sa's for the controller and other components  should help

  - Get, Update
- image.openshift.io/ImagesStream/layers (required for internal registry access)
  - Get
- operator.openshift.io/ImageContentSourcePolicy
- config.openshift.io/ImageDigestMirrorSet
- config.openshift.io/ImageTagMirrorSet
  - Watch, List, Get
- config.openshift.io/Image
  - Watch, List, Get

#### PodPlacementConfig CR
The configuration CR of the operator allows the admins to set the following:
- a `NamespaceSelector` to filter the namespaces MMO should patch the pods of.
The `openshift-*`, `kube-*` and `hypershift-*` namespaces are implicitly excluded.
- the `LogVerbosity` level of the operator
A new Custom Resource Definition (CRD) is being proposed to handle the
configuration of the operator deployment.

```go
// +kubebuilder:validation:Enum=Normal;Debug;Trace;TraceAll
type LogVerbosityLevel string

const (
    LogVerbosityLevelNormal LogVerbosityLevel = "Normal"
    LogVerbosityLevelDebug LogVerbosityLevel = "Debug"
    LogVerbosityLevelTrace LogVerbosityLevel = "Trace"
    LogVerbosityLevelTraceAll LogVerbosityLevel = "TraceAll"
)

// PodPlacementConfigSpec defines the desired state of PodPlacementConfig
type PodPlacementConfigSpec struct {

>     // LogVerbosity is the log level for the pod placement controller

Reviewer 7: Godoc should start with the json tag version of the name
```suggestion
    // logVerbosity is the log level for the pod placement controller
```

    // Valid values are: "Normal", "Debug", "Trace", "TraceAll".
    // Defaults to "Normal".
    // +optional
    // +kubebuilder:default=Normal
    LogVerbosity LogVerbosityLevel `json:"logVerbosity,omitempty"`

>     // NamespaceSelector is a metav1.LabelSelector to filter the namespaces whose
>     // pods should be patched with the scheduling gate at admission time.
>     // All the pods in the excluded namespaces will not be controlled by the operator.
>     // The webhook will always ignore pods in reserved namespaces matching the
>     // openshift-*, kube-*, and hypershift-* glob.

Reviewer 7: Nits, think it might be clearer worded this way
```suggestion
    // namespaceSelector is a metav1.LabelSelector to filter the namespaces whose
    // pods should be patched with the scheduling gate at admission time.
    // Any pod in a namespace not matched by the selector will not be controlled by the operator.
    // The webhook will always ignore pods in reserved namespaces matching the
    // openshift-*, kube-*, and hypershift-* globs.
```

    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`

}

// PodPlacementConfigStatus defines the observed state of PodPlacementConfig
type PodPlacementConfigStatus struct {

>     // Conditions represents the latest available observations of a PodPlacementConfig's current state.

Reviewer 7: There's a bunch of tags for list map keys and merge strategy missing here, we can pick those up int he API Pr though

        replicas=3
    }
    SchedulingGateMutatingWebhookSvc --> SchedulingGateMutatingWebhook

    class SchedulingGateMutatingWebhookSvc {
        <<service>>
    }

    SchedulingGateMutatingWebhookConfiguration --> SchedulingGateMutatingWebhookSvc
    Operator --> PodPlacementController : creates and reconciles
    Operator --> SchedulingGateMutatingWebhook : creates and reconciles
    Operator --> PlacementPodConfig : reconciles
    PodPlacementController --> Pod : updates (patches nodeAffinity,<br/>removes schedulingGate)
    SchedulingGateMutatingWebhook --> Pod : updates (adds schedulingGate)

```
### Open Questions
- What is the performance impact when pods are gated and inspected over many
  namespaces? This needs to be qualified. Mitigations have already been
  discussed in the other sections, but we expect others to address at
  implementation and performance evaluation time. Leader election time
  should also be considered when pods are gated and the primary controller is
  down. This would add seconds worth of delay to the scheduling. Is there
  scope to minimize this?
- Should a backoff retry mechanism with a maximum number of retries be

>   utilized in instances of image inspection failure (e.g., temporery registry

Reviewer 2: ```suggestion
  utilized in instances of image inspection failure (e.g., temporary registry
```

  unreachability, network partitioning, etc.) instead of immediately removing the scheduling gate?
- Do we need to be aware of any unique ways to access pull secrets for
  disconnected clusters?
- Is a cleanup mechanism needed to unblock pods that are scheduling gated upon
  the operator's uninstallation or the CR's deletion?
- Should the `SchedulingGateMutatingWebhook` and the `PodPlacementConfig` CR also consider a Pod Selector to filter
  pods that should be gated and inspected? This would allow for a more fine-grained control of the pods.
  The current proposal is to use a namespace selector to filter the namespaces for which the

>   operator should gate and inspect pods.

Reviewer 2: i think there needs to be a way to opt out of this behavior, though that might be as simple as "provide your own node selectors" since this EP elsewhere seems to imply if the pod already has a nodeselector, this controller won't touch it, correct?

Reviewer 2: yeah i think that's ok.  i'll resolve this thread.

    // logVerbosity is the log level for the pod placement controller
    // Valid values are: "Normal", "Debug", "Trace", "TraceAll".
    // Defaults to "Normal".
    // +optional
    // +kubebuilder:default=Normal
    LogVerbosity LogVerbosityLevel `json:"logVerbosity,omitempty"`

    // namespaceSelector is a metav1.LabelSelector to filter the namespaces whose
    // pods should be patched with the scheduling gate at admission time.
    // Any pod in a namespace not matched by the selector will not be controlled by the operator.
    // The webhook will always ignore pods in reserved namespaces matching the
    // openshift-*, kube-*, and hypershift-* globs.
    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`

}

// PodPlacementConfigStatus defines the observed state of PodPlacementConfig
type PodPlacementConfigStatus struct {
    // Conditions represents the latest available observations of a PodPlacementConfig's current state.
    Conditions []metav1.Condition `json:"conditions,omitempty"`
}

//+kubebuilder:object:root=true
//+kubebuilder:subresource:status

// PodPlacementConfig is the Schema for the podplacementconfigs API

> type PodPlacementConfig struct {

Reviewer 9: A few related questions:
- I assume this is expected to be a singleton CR since it controls the verbosity level of the controller? If so, I'd suggest including CEL-based validation for a specific `metadata.name`.
- Is this CRD cluster-scoped or namespace-scoped? I assume cluster-scoped since its configuration results in behavior changes across other namespaces?

Reviewer 9: Example of CEL-based singleton validation for a cluster-scoped object here: https://github.com/openshift/api/blob/a7ca92db022ae3ccf29cfa8865e3e35e71013de2/operator/v1alpha1/types_olm.go#L17

Author: Yes - the CR is a singleton and cluster scoped. I have updated the doc to reflect that and also added the CEL-based validation.

	// +kubebuilder:default=Normal
	LogVerbosity LogVerbosityLevel `json:"logVerbosity,omitempty"`

	// ExcludedNamespaces is the list of namespaces the pod placement controller should ignore.
	// When omitted, this means no opinion and the operator is left to choose a reasonable default.
	// The current default behaviour is to ignore only reserved namespaces.
	// Reserved namespaces are those matching the following globs; openshift-*,kube-*,hypershift-*.
	// The list of reserved namespaces will always be excluded, independent of the value of this field.

> 	// The list of reserved namespaces is subject to change over time.

Reviewer 7: Do we still want this behaviour with the namespace selector? Or are we dropping the restriction of not applying this to openshift/kube/hypershift namespaces?

Author: openshift/kube/hypershift namespaces are excluded implicitly based on the namespace name. Unfortunately there is no specific label; we can use to exclude these (I think). this is already stated I believe.

Reviewer 7: The information about excluding these was dropped from the godoc in the latest update, so we should make sure it makes its way back in

Reviewer 3: Fixed.

> 	// it never skips the policy.

Reviewer 7: Was this copied from somewhere? I'm not sure this is the correct godoc/context for the particular use case of the selector

Author: this was the default namespace selector text that I lazily copied. I'll change this to reflect this specific case.

Reviewer 3: Fixed too.

Author: ```
// By default, the labelSelector consists of a single match expression that filters out
// pods in namespaces whose the label `multiarch.openshift.io/excludeNamespace` exists.
```

@aleskandro the `multiarch.openshift.io/excludeNamespace` is just an example here. I'm not sure if we should mention it in the docs. The administrator will have the flexibility to control it. It need not necessarily be an exclusion criteria.

Reviewer 3: I removed it from the godoc

- What is the performance impact when pods are gated and inspected over many
  namespaces? This needs to be qualified. Mitigations have already been
  discussed in the other sections, but we expect others to address at
  implementation and performance evaluation time. Leader election time
  should also be considered when pods are gated and the primary controller is
  down. This would add seconds worth of delay to the scheduling. Is there

>   scope to minimize this?

Reviewer 7: If you can use `ReleaseOnCancel` then that will help to prevent long periods, but, to use `ReleaseOnCancel` you need to be sure you're shutting down correctly all reconciles with context management etc

Reviewer 3: Hi @JoelSpeed , are you concerned about the possibility of some pods being scheduled without the nodeAffinity patch due to the missing schedulingGate, about the potential delay that scheduling gated pods might experience during the scheduling process, or both?

Author: i think the concern here is more about a race with both controllers being active if  there are issues in termination: https://github.com/kubernetes/client-go/blob/master/tools/leaderelection/leaderelection.go#L153

Reviewer 3: What's the critical path in the case of the PodPlacementController?

We are configuring the PodPlacementController as active-passive for HA because it's challenging to assign shards of pods to (active-active) replicas of the controller reliably.

Using releaseOnCancel, if multiple reconciliations run concurrently on the same pod object, the Kubernetes optimistic concurrency principle ensures that only one process will succeed while the others will fail. 

The pods scheduling process will continue succeeding though.

Reviewer 7: I'm concerned that a Pod may get the gate added, and then not be removed for an extended period (compared to normal operation), because the controller is shutting down and needs time to hand over to the next pod. If you use ReleaseOnCancel, the lease handover can be quite quick, without it, it can take upwards of a minute with normal OpenShift leader configuration

Reviewer 3: Added a point discussing the releaseOnCancel option as a mitigation to the performance-related risks.
